src/test/java/com/neo4j/docker/TestUpgrade.java

```diff
@@ -4,60 +4,208 @@
-		{
-			HostFileSystemOperations.mountHostFolderAsVolume( container, dataMount, "/data" );
-			container.start();
-			DatabaseIO db = new DatabaseIO( container );
-			db.putInitialDataIntoContainer( user, password );
-		}
-
-		try(GenericContainer container = makeContainer( TestSettings.IMAGE_ID ))
-		{
-			HostFileSystemOperations.mountHostFolderAsVolume( container, dataMount, "/data" );
-			container.start();
-			DatabaseIO db = new DatabaseIO( container );
-			db.verifyDataInContainer( user, password );
-		}
-	}
-
-	// todo add test for 4.0 when I can figure out how to close a container cleanly
+    }
+
+    @Test
+    void canUpgradeFromBeforeFilePermissionFix35() throws Exception
+    {
+        Neo4jVersion beforeFix = new Neo4jVersion( 3, 5, 3 );
+        String beforeFixImage = (TestSettings.EDITION == TestSettings.Edition.ENTERPRISE) ? "neo4j:3.5.3-enterprise" : "neo4j:3.5.3";
+        Assumptions.assumeTrue( TestSettings.NEO4J_VERSION.isNewerThan( beforeFix ), "test only applicable to latest 3.5 docker" );
+        Assumptions.assumeFalse( TestSettings.NEO4J_VERSION.isAtLeastVersion( Neo4jVersion.NEO4J_VERSION_400 ),
+                                 "test only applicable to latest 3.5 docker" );
+
+        Path dataMount = HostFileSystemOperations.createTempFolder( "data-upgrade-" );
+        log.info( "created folder " + dataMount.toString() + " to test upgrade" );
+
+        try ( GenericContainer container = makeContainer( beforeFixImage ) )
+        {
+            HostFileSystemOperations.mountHostFolderAsVolume( container, dataMount, "/data" );
+            container.start();
+            DatabaseIO db = new DatabaseIO( container );
+            db.putInitialDataIntoContainer( user, password );
+        }
+
+        try ( GenericContainer container = makeContainer( TestSettings.IMAGE_ID ) )
+        {
+            HostFileSystemOperations.mountHostFolderAsVolume( container, dataMount, "/data" );
+            container.start();
+            DatabaseIO db = new DatabaseIO( container );
+            db.verifyDataInContainer( user, password );
+        }
+    }
+
+    @Test
+    void canUpgradeFromReleasedVersion() throws Exception
+    {
+        var targetNeo4jVersion = TestSettings.NEO4J_VERSION;
+
+        // If this is the very first in a new major series (i.e. a .0.0 release) then this test isn't expected to work
+        Assumptions.assumeFalse( targetNeo4jVersion.major == 0 && targetNeo4jVersion.minor == 0 );
+
+        // TODO: update this when moving to the next minor release
+        // I am taking a guess here that we will have published 4.1.0 by 1st of July
+        if ( Instant.now().isBefore( Instant.parse( "2020-07-01T00:00:00.00Z" ) ) )
+        {
+            // This sort-of-hack is necessary when we cut a new minor branch before the previous minor branch has been published to dockerhub.
+            Assumptions.assumeTrue( Neo4jVersion.NEO4J_VERSION_420.isNewerThan( targetNeo4jVersion ) );
+        }
```

Comment:
For the main Neo4j Jars we say that going from 1 minor version to another is supported.
For clustering two adjacent minor versions are required to form a functioning cluster (necessary for no-downtime upgrades).
So my intention is to test that migration to `x.y.0` from `x.y-1` works.

The problem I'm trying to tackle is that there is a period (like now) when 4.2.0 exists internally (so these tests will be run in PRs etc.) but there is no published 4.1 to test upgrade from.

Reply:
why not just do that kind of upgrade in a separate unit test rather than having over-complicated test logic here? Actually that would simplify the logic a lot since the latest patch version of a release is always tagged as the major.minor. EG `neo4j:4.0` is always the latest 4.0 patch version.

Reply:
I will separate so we will do:

1. patch upgrade tests I.e. from x.y (released) to current head (assuming current head patch version > 0)

2. Minor version upgrade test I.e. from x.(y-1) to current head.

For the latter there will still need to be a temporary hack for situations like at the moment where there is a 4.2 branch but no 4.1 release.

What’s going on should be much more transparent though